enhancements/oc/image-content-source-policy-awareness.md

```diff
@@ -0,0 +1,107 @@
+using the image-reference from the user-given image.
+
+Current bugs regarding this Issue:   
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823839
+* https://bugzilla.redhat.com/show_bug.cgi?id=1823143 and also for 4.3, 4.5, 4.6
+
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Open Questions
+
+* Should new flags be introduced OR should oc try each scenario until it succeeds in finding the registry/repo/name:digest?
+    * NO FLAGS:
+        1. Try ImageContentSources from cluster -> if oc can access ICSP and cluster, get images from ICSP, check if an image exists, if so use it
+        2. Try user-given registry/repo/name -> check if registry/repo/name:digest exists, if so use it
+        3. (The current flow): If the above fall through, try the image-reference from user-given image - use it - may or may not succeed 
+        if disconnected or not authorized to access image-reference registry
+    * FLAGS:
+        * *--icsp-to* will define where to write an ICSP file to.  If unset, `oc adm release mirror` will write to current directory.
+        * *--icsp-file* will define where to get ICSP from a file.  If set, `oc adm release extract|mirror` will use this ICSP data.
+    * FLAGS DECIDED AGAINST:
+        * boolean `--use-icsp` and if true, check for cluster and/or icsp file?  This is problematic, because even if a user is currently connected to a cluster, 
+        it doesn't mean they want to use information from that cluster with an `oc adm release ...` command.  
+        So here, we'd need a different `--cluster-icsp` boolean and `--icsp-file` string flag.  The flags are adding up here, and that is not desireable. 
+        * string `--image-content-source aregistry/arepo/arelease`. I don't like this, because it would be redundant for a user to run something like this: 
+        `oc adm release extract --command oc myreg:5000/myrepo/release:tag --image-content-source myreg:5000/myrepo/release`.  
+        * Instead, a boolean `--set-prefix` would allow a user to specify "I want to use the prefix of the release image I have specified, 
+        rather than any underlying image reference."
+* This proposal is for the `FLAGS`, that `oc` will try ICSP, user-given image, then fall back on underlying image-reference from user-given image. 
+* See `User Stories` below for examples. 
+
+## Summary
+
+ICSP allows OpenShift (CVO, CRI-O) to check down a list of possible mirrors to find an image with the matching digest it is
+looking for.  `oc` should do the same.  If no ICSP found, then try user-given image.  If that doesn't succeed, then use underlying
```

Comment:
`oc` should do the same because users expect that when they interact with a cluster, the cluster's context informs their action (i.e. `oc adm release info` will lookup the current cluster).  In general, letting `oc` default to the behavior of the connected cluster is the right thing to do when the current cluster is the target - but not when the context is unclear or may be a different target.

This may help clarify this summary.

Reply:
Clarifications I'd like added to the proposal (maybe as "design points") for future reference:

1. `oc` can use context of the cluster when the action the user is taking is clearly expected to talk to the cluster (`oc get foo`)
2. When the user's action is not obviously expected to talk to the cluster (`oc adm release info foo`), the cli MAY attempt to load ICSP info from an existing clusters, but if that fails, the cli MUST ignore that failure (log at a high debug level)
3. In the case where the CLI attempts to lookup the ICSP in order to help the user, that should happen after the first attempt to retrieve content from the location fails, in which case the ICSP should be looked up and an attempt made to find the alternate location from those sources - if all of those fail, then the original error must be returned.
4. In general, when a user specifies an explicit ICSP, the CLI MUST fail if that ICSP cannot be loaded, and SHOULD follow the order defined in the ICSP

Reply:
added